app/src/main/java/org/wikipedia/page/leadimages/LeadImagesHandler.java

```diff
@@ -134,40 +147,60 @@ private void updateCallToAction() {
-                                break;
+                        ServiceFactory.get(new WikiSite(Service.COMMONS_URL)).getImageInfo(imageTitle, WikipediaApp.getInstance().getAppOrSystemLanguageCode()), Pair::new))
+                .flatMap(pair -> {
+                            captionSourcePageTitle = new PageTitle(imageTitle, new WikiSite(Service.COMMONS_URL, getTitle().getWikiSite().languageCode()));
+                            imageCaptions = pair.first;
+                            imageInfo = pair.second.query().firstPage().imageInfo();
+                            imagePage = pair.second.query().firstPage();
+                            if (!imageCaptions.containsKey(getTitle().getWikiSite().languageCode())) {
+                                imageEditType = ImageEditType.ADD_CAPTION;
+                                return ImageTagsProvider.getImageTagsObservable(pair.second.query().firstPage().pageId(), getTitle().getWikiSite().languageCode());
+                            }
+                            if (app.language().getAppLanguageCodes().size() >= MIN_LANGUAGES_TO_UNLOCK_TRANSLATION) {
+                                for (String lang : app.language().getAppLanguageCodes()) {
+                                    if (!imageCaptions.containsKey(lang)) {
+                                        imageEditType = ImageEditType.ADD_CAPTION_TRANSLATION;
+                                        captionTargetPageTitle = new PageTitle(imageTitle, new WikiSite(Service.COMMONS_URL, lang));
+                                        break;
+                                    }
+                                }
                             }
+                            return ImageTagsProvider.getImageTagsObservable(pair.second.query().firstPage().pageId(), getTitle().getWikiSite().languageCode());
                         }
+                )
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(imageTagsResult -> {
+                    if (imageEditType != ImageEditType.ADD_CAPTION && imageTagsResult != null && imageTagsResult.size() == 0) {
+                        imageEditType = ImageEditType.ADD_TAGS;
                     }
-                }, L::e));
+                    updateCta();
+                })
+        );
+    }
+
+    private void updateCta() {
+        switch (imageEditType) {
+            case ADD_TAGS:
+                pageHeaderView.setUpCallToAction(app.getResources().getString(R.string.suggested_edits_article_cta_image_tags));
+                break;
+            case ADD_CAPTION_TRANSLATION:
+                callToActionIsTranslation = true;
+                String currentCaption = imageCaptions.get(getTitle().getWikiSite().languageCode());
+                captionSourcePageTitle.setDescription(currentCaption);
```

Comment:
Would it be better if set the caption in the `flatMap` in line `151`?

Reply:
Cooltey this was just convenient as we use it only here... but if there is any other use please lmk and  i will move it

Reply:
It would be good if we can reduce one class-level variable.